### ISSUE TITLE
Status filter is not working

### ISSUE THREAD

**Issue by stellarmagnet** [AutarkLabs/open-enterprise#1488]
https://rinkeby.aragon.org/#/oe-demo/0xf1c7d9fea39b3fae66358f0b595b175a03828ccc

Try the status filter -- you will see it is not properly filtering. This should be fully tested for all bounty statuses and improved. Please be thorough in the testing of all of the filters.

![image](https://user-images.githubusercontent.com/2584493/67801835-43e1bc00-fa47-11e9-8b5d-501480fb87a8.png)

**Comment by chadoh** [comment]
@rkzel and I looked into this together, and here's what we discovered.

![confusing filter counts](https://user-images.githubusercontent.com/221614/67869943-8a492080-fb04-11e9-89ef-84f88551417a.gif)

---

As seen in the GIF above, the counts correspond to issues-with-bounties from _any_ project, not just the currently-selected project.

Note, however, that this will go beyond _currently-selected_projects_. If you are looking at all _projects_, but filtering by a particular _label_, we probably want the count of "Accepting Applicants" to reflect only issues with that label that are also accepting applicants.

A *quick solution*: we can make these counts respect the other filters that have been applied. However, there is an additional UX edge case that this will not cover:

![when issues with funding haven't been loaded yet](https://user-images.githubusercontent.com/221614/67871581-fb89d300-fb06-11e9-843c-d5fca0b5aff2.gif)

---

Above, we see a situation where the issue with funding _does_ match the given filter set, but the first page of 100 results from the initial query did not contain this issue.

My suggestion for this situation: keep the count showing values for not-yet-loaded issues, if those issues match the other filters. Then, when that filter is selected, automatically fire the query to load matching issues, rather than showing an empty list and "load more".

Unfortunately, this is a *complex solution*. Our code is not currently written in a way that makes this straightforward. I can spend some more time thinking through it; maybe there's a way to achieve this behavior in just a day or two, but I suspect to do it well will require some significant refactoring + associated QA efforts, totaling about a week of effort.

---

One last consideration: the *quick solution* (making the counts show the number of _downloaded_ matching issues) conflicts with the *complex solution.* In the second screen recording above, imagine the count only showing the number of _downloaded_ matching issues. It would show `0`, rather than `1`. Do we want that?

**Comment by chadoh** [comment]
# Technical task breakdown, with built-in UX trade-off assumptions

We want it to work like this:

<img width="1206" alt="project filtering logic" src="https://user-images.githubusercontent.com/221614/69277572-7eee8f80-0bae-11ea-8d24-c909d5304aea.png">

Walking through an example:

If we select one kind of label ("Help Wanted," say), then we should get _only_ issues with that label. If we then select another label ("Good First Issue"), we should get _more_ issues; all issues with _either_ label.

If we further filter by milestone ("Q1 2020," say), we should get _only_ issues in that milestone which _also_ match the set of selected labels.

Furthermore, we also want the counts next to each filter to match the _total number of matched issues on GitHub_. That is, if there are 50 issues in the "Q1 2020" milestone, but only 5 that have either the "Help Wanted" or "Good First Issue" label, then the filter text should appear as "Q1 2020 (5)".

An example of how we want this to work: [newegg's filtering](https://www.newegg.com/All-in-One-Computers/SubCategory/ID-3309?Tid=167567).

Here's how we can approach this:

## Part 1: Re-querying GitHub when user selects a new filter

- [ ] Remove all counts next to filters; getting it to work right is tricky (see Part 2 below)
- [ ] On initial app load, query GitHub for all Milestones and Labels for all Projects; store this in Filters global context (#1673); use this to build FilterBar, rather than currently-downloaded issues. Since filter counts are not shown, this state will stay constant, even if applying further filters would not make sense (example: there are zero issues with the selected labels in the "Q1 2020" milestone – the "Q1 2020" milestone would still be shown, and selecting it would show an empty list)
- [ ] Move querying logic itself to Filters global context; construct GitHub queries based on selected filters; fire new queries as new filters are selected; remove frontend-filtering logic

How long will this take? **Estimate: 5 days**. These individual items could be worked in parallel by multiple engineers, reducing total time somewhat.

## Part 2: Updating filter counts based on current filters

- [ ] To make counts next to filter options work correctly, when filtering by –
  * Project:
    * query GitHub for Labels for this project & issues count for each
    * query GitHub for Milestones for this project & issues count for each
    * on frontend, loop over bountyIssues, update counts according to bountyIssues relevant to this project
  * Label:
    * query GitHub for Milestones applicable to issues with this label & issues count for each
    * leave existing Labels counts as-is
    * on frontend, loop over bountyIssues, update counts according to bountyIssues relevant to this Label
  * Milestone:
    * same idea as Label (but swapping logic for Milestones & Labels, obvs)
  * Status
    * query GitHub for the specific set of issues with these IDs
    * on frontend, update counts for Labels and Milestones according to the selected issues
- [ ] Rebuild list of available filters every time a filter is selected, since it's not useful to allow applying a filter that would show zero issues

How long will this take? **Estimate: 5 days**. These items must be worked sequentially, and getting it right is going to be tricky. Making all of this code clean/understandable/maintainable is going to add additional time.

## Part 3: Improve sorting; allow sorting by "Date Funded"

Unrelated to "Date Funded" sorting:

- [ ] When selecting a new sort-by option, re-query GitHub, build a new list of issues (right now we attempt to only sort the issues that have been loaded into the browser so far, and we don't do it very well)

^ this could be done at any time, and is mostly unrelated to all the other work here.

Sorting by "Date Funded," however, should probably be done at least after the Part 1 outlined above, as this new sorting option will further complicate our filtering logic, and Part 1 gets filtering working acceptably.

To implement sorting by "Date funded":

- [ ] On initial load of Projects, while still on Overview tab, load all issues-with-funding in the background, similar to how we do with `DecoratedRepos`. We can query GitHub for `nodes` and give it a list of ids; we will need to store ids (not numbers) for every funded issue.
- [ ] On Issues tab, add new "Sort By" option, called "Date funded descending". Make this the default sort. Also add "Date funded ascending", for extra credit. Sorting by either of these can immediately show the full set of issues-with-funding, which is already stored in the user's browser. These will all be listed (and sorted correctly) at the top of the list, and then below this, we can show the "loading" animation while we fetch 100 more issues from GitHub. We can explore the possibility of instructing GitHub to omit the issues-with-funding; if it does not support such a query, then we can filter out issues-with-funding on the frontend.
- [ ] When sorting other ways, we can ignore this set of funded issues, and stop omitting them from the main list. This way they'll fall into place in the list as they do now.

How long will this take? **Estimate: 5 days**.

---

Something else to consider: the above intertwines our logic more and more with what GitHub allows. The more complicated our GitHub logic, the more difficult we will find extricating ourselves from GitHub as the sole Issue Provider for the Projects app. We may want to take a detour through building support for another Issue Provider, and only then return to filtering.

---

@stellarmagnet @jayalaves what do you think of

1. Hiding filter counts for now
2. Skipping all of this for now, leaving filtering as-is (which is, unfortunately, quite broken), and only returning to it after we support a new Issue Provider

**Comment by ottodevs** [comment]
I vote for the filtered queries solution, is the most professional one and the most adequate for this use case. I think that is possibly the solution you are proposing @chadoh querying ~GitHub~Apollo for results after applying a filter, and Apollo will decide if that query needs to hit GitHub or can be served from the cache?

**Comment by jayalaves** [comment]
@chadoh Your proposal sounds like an elegant solution that would effectively improve the user experience. As you've pointed out, our intention is to facilitate GitHub-independent functionality on the Projects app, including the ability, but not limited to:

* Creating, editing and deleting repositories
* Creating, editing and deleting issues, whereby issues can be composed in similar fashion to how you'd be able to compose an issue on GitHub, with Markdown support, the ability to add labels, or associate them to a milestone
* Defining a simple set of project management features, including but not limited to assigning a member of your organization to work on an issue (which is not the same as assigning a bounty), having a basic review process, or facilitating kanban views associated to progress tags

And on top of this standalone set of features, we then should think about what the "Connect to GitHub" experience looks like, since some users may still find value in connecting their GitHub organizations to an Aragon DAO, so how do we combine the two systems effectively? This is a design challenge for which I personally don't have a definite answer for at the moment.

"Enhancing the Projects app" was a value proposition of its own as part of AGP-73, and much of it was associated to decoupling from GitHub as a necessary piece of the puzzle. So, if the Projects app no longer needs GitHub, then what _is_ the Projects app? Is it a full replacement for GitHub or is it something else? Should it be more like Trello than GitHub? Are complex filters an integral part of that new user experience, or are they not?

**Comment by rkzel** [comment]
> So, if the Projects app no longer needs GitHub, then what is the Projects app?

Projects App connects value to tasks in form of bounties. We're in software world  right now, so GH was obvious choice for being an Issue Provider. However, it was never intended to be the sole IP - interfacing with other IPs was the next step, as was being an IP ourselves.

> Is it a full replacement for GitHub or is it something else?

If you want to give someone 1 ETH for finding you a perfect cheesecake recipe you probably need neither GH nor its 1:1 replacement (not a perfect example though, version control for recipes seems like a good idea).

> Should it be more like Trello than GitHub?

Based on my experience with quite a few IPs - yeah, more Trello/Asana/ZenHub than plain GH. 

> Are complex filters an integral part of that new user experience, or are they not?

If there is a thousand issues/tasks (whatever we choose to name them), there must be an easy way to find/manage tasks. Complex filtering is therefore an integral part of what needs building.

A task is... um, an entity ;) in its minimal form consisting of title, body and bounty, having a list of connections to people (who created, applied, worked on, got bounty) from which history can be derived. Everything else are means to quickly find what one is looking for - and provide quick overview of the whole. So, tasks have attributes like lablels, milestones, priorities, queues, points, weights and whatever other attribute we can think of that serves one basic purpose: to know what is there.

So, there is a basic text search (looking at title), and then a set of quite similar filters operating in similar manner, one for each attribute, and then some additional nice-to-haves filters looking at history (find tasks completed by X, find tasks completed in January), etc...

GitHub is software-oriented. We want something simpler and more universal, not 1:1 replacement, but complex filtering is unavoidable.

**Comment by javieralaves** [comment]
Iceboxing this for now until it is later addressed by higher level changes in the Projects app.
